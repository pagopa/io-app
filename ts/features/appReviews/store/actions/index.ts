import { ActionType, createStandardAction } from "typesafe-actions";
import { AppFeedbackUri } from "../../../../../definitions/content/AppFeedbackUri";

export type TopicKeys = keyof AppFeedbackUri;

export const appReviewPositiveFeedback = createStandardAction(
  "APP_REVIEW_POSITIVE_FEEDBACK"
)();

export const appReviewNegativeFeedback = createStandardAction(
  "APP_REVIEW_NEGATIVE_FEEDBACK"
)<TopicKeys>();

// Reset feedback data just for testing purposes
export const clearFeedbackDatas = createStandardAction(
  "APP_REVIEW_RESET_FEEDBACK"
)();

export type AppFeedbackActions =
  | ActionType<typeof appReviewPositiveFeedback>
  | ActionType<typeof appReviewNegativeFeedback>
  | ActionType<typeof clearFeedbackDatas>;
