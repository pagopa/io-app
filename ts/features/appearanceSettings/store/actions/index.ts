/**
 * Action types and action creator related to the Appearance settings.
 */
import { ActionType, createStandardAction } from "typesafe-actions";

export const setShowAppearanceSettingsBanner = createStandardAction(
  "SHOW_APPEARANCE_SETTINGS_BANNER"
)<boolean>();

export type AppearanceSettingsActions = ActionType<
  typeof setShowAppearanceSettingsBanner
>;
