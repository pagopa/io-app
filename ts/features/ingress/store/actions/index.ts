import { ActionType, createStandardAction } from "typesafe-actions";

export const setIsBlockingScreen = createStandardAction(
  "SET_IS_BLOCKING_SCREEN"
)();

export type IngressScreenActions = ActionType<typeof setIsBlockingScreen>;
