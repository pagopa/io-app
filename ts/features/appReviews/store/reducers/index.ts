import { getType } from "typesafe-actions";
import { PersistConfig, persistReducer } from "redux-persist";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { Action } from "../../../../store/actions/types";
import {
  appReviewNegativeFeedback,
  appReviewPositiveFeedback,
  clearFeedbackDatas,
  TopicKeys
} from "../actions";
import {
  logoutFailure,
  logoutSuccess
} from "../../../../store/actions/authentication";

// String formats for the dates represents the log to be persisted to check the last feedback request time
// both fields are updated directly from reducer on action dispatch and not from the action payload.
export type AppFeedbackState = {
  positiveFeedbackDate: string;
  negativeFeedbackDate:
    | Record<TopicKeys, string | undefined>
    | Record<string, never>;
};

export const appFeedbackInitialState: AppFeedbackState = {
  positiveFeedbackDate: "",
  negativeFeedbackDate: {}
};

const appFeedbackReducer = (
  state: AppFeedbackState = appFeedbackInitialState,
  action: Action
): AppFeedbackState => {
  switch (action.type) {
    case getType(logoutSuccess):
    case getType(logoutFailure):
    case getType(clearFeedbackDatas):
      return {
        ...state,
        ...appFeedbackInitialState
      };
    case getType(appReviewPositiveFeedback):
      return {
        ...state,
        positiveFeedbackDate: new Date().toISOString()
      };
    case getType(appReviewNegativeFeedback):
      return {
        ...state,
        negativeFeedbackDate: {
          ...state.negativeFeedbackDate,
          [action.payload]: new Date().toISOString()
        } as Record<TopicKeys, string | undefined>
      };
    default:
      return state;
  }
};

export const persistConfig: PersistConfig = {
  key: "appFeedback",
  storage: AsyncStorage
};

export const appFeedbackPersistor = persistReducer<AppFeedbackState, Action>(
  persistConfig,
  appFeedbackReducer
);
