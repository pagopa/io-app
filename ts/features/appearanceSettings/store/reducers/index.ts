import AsyncStorage from "@react-native-async-storage/async-storage";
import { persistReducer, PersistConfig } from "redux-persist";
import { getType } from "typesafe-actions";
import { differentProfileLoggedIn } from "../../../../store/actions/crossSessions";
import { Action } from "../../../../store/actions/types";
import { setShowAppearanceSettingsBanner } from "../actions";

export type AppearanceSettingsState = {
  showAppearanceBanner: boolean;
};

export const appearanceSettingsReducerInitialState = {
  showAppearanceBanner: true
};

const appearanceSettingsReducer = (
  state: AppearanceSettingsState = appearanceSettingsReducerInitialState,
  action: Action
): AppearanceSettingsState => {
  switch (action.type) {
    case getType(setShowAppearanceSettingsBanner): {
      return {
        ...state,
        showAppearanceBanner: action.payload
      };
    }
    case getType(differentProfileLoggedIn): {
      return appearanceSettingsReducerInitialState;
    }
    default:
      return state;
  }
};

const persistConfig: PersistConfig = {
  key: "appearanceSettings",
  storage: AsyncStorage
};

export const appearanceSettingsReducerPersistor = persistReducer(
  persistConfig,
  appearanceSettingsReducer
);
