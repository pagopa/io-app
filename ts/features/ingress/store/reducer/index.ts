import { getType } from "typesafe-actions";
import { setIsBlockingScreen } from "../actions";
import { Action } from "../../../../store/actions/types";

export type IngressScreenState = {
  isBlockingScreen: boolean;
};

export const initialIngressScreenState: IngressScreenState = {
  isBlockingScreen: false
};

export const ingressScreenReducer = (
  state = initialIngressScreenState,
  action: Action
): IngressScreenState => {
  switch (action.type) {
    case getType(setIsBlockingScreen):
      return {
        ...state,
        isBlockingScreen: true
      };
    default:
      return state;
  }
};
