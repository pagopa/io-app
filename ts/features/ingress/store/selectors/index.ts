import { GlobalState } from "../../../../store/reducers/types";

export const isBlockingScreenSelector = (state: GlobalState) =>
  state.features.ingress.isBlockingScreen;
