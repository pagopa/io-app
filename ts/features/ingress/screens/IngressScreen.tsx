/* eslint-disable functional/immutable-data */
/**
 * An ingress screen to choose the real first screen the user must navigate to.
 */
import React, { memo, useEffect, useRef, useState } from "react";
import {
  fetch as fetchNetInfo,
  NetInfoState
} from "@react-native-community/netinfo";
import { Millisecond } from "@pagopa/ts-commons/lib/units";
import { AccessibilityInfo, View } from "react-native";
import I18n from "../../../i18n";
import { isMixpanelEnabled as isMixpanelEnabledSelector } from "../../../store/reducers/persistedPreferences";
import { trackIngressScreen } from "../../../screens/profile/analytics";
import LoadingScreenContent from "../../../components/screens/LoadingScreenContent";
import { OperationResultScreenContent } from "../../../components/screens/OperationResultScreenContent";
import { useIODispatch, useIOSelector } from "../../../store/hooks";
import { isBackendStatusLoadedSelector } from "../../../store/reducers/backendStatus";
import { setIsBlockingScreen } from "../store/actions";
import ModalSectionStatusComponent from "../../../components/SectionStatus/modal";
import { isMixpanelInitializedSelector } from "../../mixpanel/store/selectors";
import {
  trackIngressCdnSystemError,
  trackIngressServicesSlowDown,
  trackIngressTimeout
} from "../analytics";
import { setAccessibilityFocus } from "../../../utils/accessibility";

const TIMEOUT_CHANGE_LABEL = (5 * 1000) as Millisecond;
const TIMEOUT_BLOCKING_SCREEN = (10 * 1000) as Millisecond;

export const IngressScreen = () => {
  const isMixpanelInitialized = useIOSelector(isMixpanelInitializedSelector);
  const isMixpanelEnabled = useIOSelector(isMixpanelEnabledSelector);
  const dispatch = useIODispatch();
  const [netInfo, setNetInfo] = useState<NetInfoState>();
  const [showBlockingScreen, setShowBlockingScreen] = useState(false);
  const [contentTitle, setContentTitle] = useState(I18n.t("startup.title"));

  useEffect(() => {
    // Since the screen is shown for a very short time,
    // we prefer to announce the content to the screen reader,
    // instead of focusing the first element.
    AccessibilityInfo.announceForAccessibilityWithOptions(contentTitle, {
      queue: true
    });
  }, [contentTitle]);

  useEffect(() => {
    // `isMixpanelEnabled` mustn't be `false`
    if (isMixpanelInitialized && isMixpanelEnabled !== false) {
      trackIngressScreen();
    }
  }, [isMixpanelInitialized, isMixpanelEnabled]);

  useEffect(() => {
    const timeouts: Array<number> = [];
    timeouts.push(
      setTimeout(() => {
        setContentTitle(I18n.t("startup.title2"));
        timeouts.shift();
      }, TIMEOUT_CHANGE_LABEL)
    );
    timeouts.push(
      setTimeout(() => {
        setShowBlockingScreen(true);
        dispatch(setIsBlockingScreen());
        timeouts.shift();
      }, TIMEOUT_BLOCKING_SCREEN)
    );
    void fetchNetInfo()
      .then(info => {
        if (!info.isConnected) {
          timeouts.forEach(clearTimeout);
        }
        setNetInfo(info);
      })
      .catch();

    return () => {
      timeouts?.forEach(clearTimeout);
    };
  }, [dispatch]);

  if (netInfo && !netInfo.isConnected) {
    return (
      <OperationResultScreenContent
        testID="device-connection-lost-id"
        pictogram="lostConnection"
        title={I18n.t("startup.connection_lost.title")}
        subtitle={I18n.t("startup.connection_lost.description")}
      />
    );
  }

  if (showBlockingScreen) {
    return <IngressScreenBlockingError />;
  }

  return (
    <>
      <ModalSectionStatusComponent
        sectionKey="ingress"
        sticky
        trackingAction={trackIngressServicesSlowDown}
      />
      <LoadingScreenContent
        testID="ingress-screen-loader-id"
        contentTitle={contentTitle}
      />
    </>
  );
};

const IngressScreenBlockingError = memo(() => {
  const operationRef = useRef<View>(null);
  const isBackendStatusLoaded = useIOSelector(isBackendStatusLoadedSelector);
  const isMixpanelEnabled = useIOSelector(isMixpanelEnabledSelector);

  useEffect(() => {
    setAccessibilityFocus(operationRef);
  }, []);

  useEffect(() => {
    // It's not necessary to check if mixpanel is initialized since this screen is shown after 10 seconds.
    // If mixpanel is not initialized at that time, we have an issue spanning, system-wide.
    if (isMixpanelEnabled !== false) {
      if (isBackendStatusLoaded) {
        void trackIngressTimeout();
      } else {
        void trackIngressCdnSystemError();
      }
    }
  }, [isBackendStatusLoaded, isMixpanelEnabled]);

  return (
    <OperationResultScreenContent
      ref={operationRef}
      testID="device-blocking-screen-id"
      {...(isBackendStatusLoaded
        ? {
            pictogram: "time",
            title: I18n.t("startup.slowdowns_results_screen.title"),
            subtitle: I18n.t("startup.slowdowns_results_screen.subtitle")
          }
        : {
            pictogram: "umbrellaNew",
            title: I18n.t("startup.cdn_unreachable_results_screen.title"),
            subtitle: I18n.t("startup.cdn_unreachable_results_screen.subtitle")
          })}
    />
  );
});
