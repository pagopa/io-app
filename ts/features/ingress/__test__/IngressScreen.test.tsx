/* eslint-disable @typescript-eslint/ban-ts-comment */
import { createStore } from "redux";
import * as netInfo from "@react-native-community/netinfo";
import { waitFor } from "@testing-library/react-native";
import { act } from "react-test-renderer";
import { IngressScreen } from "../screens/IngressScreen";
import { renderScreenWithNavigationStoreContext } from "../../../utils/testWrapper";
import { applicationChangeState } from "../../../store/actions/application";
import { appReducer } from "../../../store/reducers";
import I18n from "../../../i18n";
import * as backendStatusSelectors from "../../../store/reducers/backendStatus";

jest.useFakeTimers();

describe(IngressScreen, () => {
  it("Should be defined", () => {
    const component = renderComponent();

    expect(component).toBeDefined();
  });
  it("Should match the snapshot", () => {
    const component = renderComponent();

    expect(component).toMatchSnapshot();
  });
  describe("IngressScreen with device connection enabled", () => {
    it("Should not display OperationResultsScreenContent", async () => {
      const { queryByTestId, getByTestId } = renderComponent();
      const operationResults = await waitFor(
        () => queryByTestId("device-connection-lost-id"),
        { timeout: 1000 }
      );
      const loader = getByTestId("ingress-screen-loader-id");
      expect(operationResults).toBeNull();
      expect(loader).toBeTruthy();
    });
  });
  describe("IngressScreen with device connection disabled", () => {
    it("Should display OperationResultsScreenContent", async () => {
      jest
        .spyOn(netInfo, "fetch")
        // @ts-ignore
        .mockResolvedValueOnce({ isConnected: false });
      const { findByTestId, queryByTestId } = renderComponent();
      // Before the effect is resolved the loader should be displayed
      expect(queryByTestId("ingress-screen-loader-id")).toBeTruthy();

      const operationResults = await findByTestId("device-connection-lost-id");

      // Once the effect is resolved the loader shouldn't be displayed
      expect(queryByTestId("ingress-screen-loader-id")).toBeNull();
      expect(operationResults).toBeTruthy();
    });
  });
  describe("IngressScreen when slowdowns occur", () => {
    beforeEach(() => {
      jest.clearAllTimers();
      jest.clearAllMocks();
    });
    it("Should update LoadingScreenContent contentTitle after 5 sec and display the cdn unreachable blocking screen after 10", async () => {
      const {
        getDeviceBlockingScreen,
        queryByText,
        getFirstText,
        getSecondText
      } = await renderComponentWithSlowdowns();

      await act(() => {
        jest.advanceTimersByTime(5000);
      });

      expect(getDeviceBlockingScreen()).toBeTruthy();
      expect(
        queryByText(I18n.t("startup.cdn_unreachable_results_screen.title"))
      ).toBeTruthy();
      expect(
        queryByText(I18n.t("startup.slowdowns_results_screen.title"))
      ).toBeNull();
      expect(getFirstText()).toBeNull();
      expect(getSecondText()).toBeNull();
    });
    it("Should update LoadingScreenContent contentTitle after 5s and display the slowdowns blocking screen after 10s", async () => {
      jest
        .spyOn(backendStatusSelectors, "isBackendStatusLoadedSelector")
        .mockImplementation(() => true);

      const {
        getDeviceBlockingScreen,
        queryByText,
        getFirstText,
        getSecondText
      } = await renderComponentWithSlowdowns();

      await act(() => {
        jest.advanceTimersByTime(5000);
      });

      expect(getDeviceBlockingScreen()).toBeTruthy();
      expect(
        queryByText(I18n.t("startup.cdn_unreachable_results_screen.title"))
      ).toBeNull();
      expect(
        queryByText(I18n.t("startup.slowdowns_results_screen.title"))
      ).toBeTruthy();
      expect(getFirstText()).toBeNull();
      expect(getSecondText()).toBeNull();
    });
  });
});

const renderComponentWithSlowdowns = async () => {
  const { queryByText, queryByTestId } = renderComponent();
  const getFirstText = () => queryByText(I18n.t("startup.title"));
  const getSecondText = () => queryByText(I18n.t("startup.title2"));
  const getDeviceBlockingScreen = () =>
    queryByTestId("device-blocking-screen-id");

  expect(getFirstText()).toBeTruthy();
  expect(getSecondText()).toBeNull();

  await act(() => {
    jest.advanceTimersByTime(5000);
  });

  expect(getFirstText()).toBeNull();
  expect(getSecondText()).toBeTruthy();
  expect(getDeviceBlockingScreen()).toBeNull();

  return {
    queryByText,
    queryByTestId,
    getFirstText,
    getSecondText,
    getDeviceBlockingScreen
  };
};

const renderComponent = () => {
  const initialState = appReducer(undefined, applicationChangeState("active"));
  const store = createStore(appReducer, initialState as any);

  return renderScreenWithNavigationStoreContext(
    IngressScreen,
    "NO_ROUTE",
    {},
    store
  );
};
