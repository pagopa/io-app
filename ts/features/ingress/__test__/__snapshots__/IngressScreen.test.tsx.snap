// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IngressScreen Should match the snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "#FFFFFF",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  NO_ROUTE
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          onGestureCancel={[Function]}
          pointerEvents="box-none"
          sheetAllowedDetents="large"
          sheetCornerRadius={-1}
          sheetExpandsWhenScrolledToEdge={true}
          sheetGrabberVisible={false}
          sheetLargestUndimmedDetent="all"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                handlerTag={2}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "#FFFFFF",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <RNCSafeAreaView
                        edges={
                          {
                            "bottom": "additive",
                            "left": "additive",
                            "right": "additive",
                            "top": "additive",
                          }
                        }
                        style={
                          {
                            "backgroundColor": "#FFFFFF",
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                        testID="ingress-screen-loader-id"
                      >
                        <View
                          style={
                            {
                              "paddingHorizontal": 24,
                            }
                          }
                        >
                          <View
                            style={
                              {
                                "alignItems": "center",
                              }
                            }
                          >
                            <View
                              accessibilityElementsHidden={true}
                              accessible={false}
                              importantForAccessibility="no-hide-descendants"
                            >
                              <View
                                accessibilityHint="Wait for the content load"
                                accessibilityLabel="Loading"
                                accessibilityRole="progressbar"
                                accessible={true}
                                importantForAccessibility="no-hide-descendants"
                                style={
                                  {
                                    "height": 48,
                                    "width": 48,
                                  }
                                }
                                testID="LoadingIndicator"
                              >
                                <View
                                  collapsable={false}
                                  style={
                                    {
                                      "transform": [
                                        {
                                          "rotateZ": "0deg",
                                        },
                                      ],
                                    }
                                  }
                                  testID="LoadingSpinnerAnimatedTestID"
                                >
                                  <RNSVGSvgView
                                    align="xMidYMid"
                                    bbHeight={48}
                                    bbWidth={48}
                                    fill="none"
                                    focusable={false}
                                    height={48}
                                    meetOrSlice={0}
                                    minX={0}
                                    minY={0}
                                    style={
                                      [
                                        {
                                          "backgroundColor": "transparent",
                                          "borderWidth": 0,
                                        },
                                        {
                                          "flex": 0,
                                          "height": 48,
                                          "width": 48,
                                        },
                                      ]
                                    }
                                    vbHeight={48}
                                    vbWidth={48}
                                    width={48}
                                  >
                                    <RNSVGGroup
                                      fill={null}
                                      propList={
                                        [
                                          "fill",
                                        ]
                                      }
                                    >
                                      <RNSVGDefs>
                                        <RNSVGLinearGradient
                                          gradient={
                                            [
                                              0,
                                              736995,
                                              1,
                                              -16040221,
                                            ]
                                          }
                                          gradientTransform={null}
                                          gradientUnits={0}
                                          name="spinner-secondHalf"
                                          x1="0%"
                                          x2="100%"
                                          y1="0%"
                                          y2="0%"
                                        />
                                        <RNSVGLinearGradient
                                          gradient={
                                            [
                                              0,
                                              -16040221,
                                              1,
                                              -16040221,
                                            ]
                                          }
                                          gradientTransform={null}
                                          gradientUnits={0}
                                          name="spinner-firstHalf"
                                          x1="0%"
                                          x2="100%"
                                          y1="0%"
                                          y2="0%"
                                        />
                                      </RNSVGDefs>
                                      <RNSVGGroup
                                        fill={
                                          {
                                            "payload": 4278190080,
                                            "type": 0,
                                          }
                                        }
                                        propList={
                                          [
                                            "strokeWidth",
                                          ]
                                        }
                                        strokeWidth={5}
                                      >
                                        <RNSVGPath
                                          d="M 2.5 24 A 21.5 21.5 0 0 1 45.5 24"
                                          fill={
                                            {
                                              "payload": 4278190080,
                                              "type": 0,
                                            }
                                          }
                                          propList={
                                            [
                                              "stroke",
                                            ]
                                          }
                                          stroke={
                                            {
                                              "brushRef": "spinner-secondHalf",
                                              "type": 1,
                                            }
                                          }
                                        />
                                        <RNSVGPath
                                          d="M 45.5 24 A 21.5 21.5 0 0 1 2.5 24"
                                          fill={
                                            {
                                              "payload": 4278190080,
                                              "type": 0,
                                            }
                                          }
                                          propList={
                                            [
                                              "stroke",
                                            ]
                                          }
                                          stroke={
                                            {
                                              "brushRef": "spinner-firstHalf",
                                              "type": 1,
                                            }
                                          }
                                        />
                                        <RNSVGPath
                                          d="M 2.5 24 A 21.5 21.5 0 0 1 2.5 22.75"
                                          fill={
                                            {
                                              "payload": 4278190080,
                                              "type": 0,
                                            }
                                          }
                                          propList={
                                            [
                                              "stroke",
                                              "strokeLinecap",
                                            ]
                                          }
                                          stroke={
                                            {
                                              "payload": 4278927075,
                                              "type": 0,
                                            }
                                          }
                                          strokeLinecap={1}
                                        />
                                      </RNSVGGroup>
                                    </RNSVGGroup>
                                  </RNSVGSvgView>
                                </View>
                              </View>
                            </View>
                            <View
                              style={
                                {
                                  "height": 24,
                                }
                              }
                            />
                            <Text
                              accessibilityLabel="We are getting ready for your experience on IO"
                              allowFontScaling={false}
                              color="bluegreyDark"
                              defaultColor="bluegreyDark"
                              defaultWeight="Semibold"
                              dynamicTypeRamp="title2"
                              font="TitilliumSansPro"
                              fontStyle={
                                {
                                  "fontSize": 22,
                                  "lineHeight": 33,
                                }
                              }
                              maxFontSizeMultiplier={1.25}
                              style={
                                [
                                  {
                                    "textAlign": "center",
                                  },
                                  {
                                    "fontSize": 22,
                                    "lineHeight": 33,
                                  },
                                  {
                                    "color": "#17324D",
                                    "fontFamily": "Titillium Sans Pro",
                                    "fontStyle": "normal",
                                    "fontWeight": "600",
                                  },
                                ]
                              }
                              weight="Semibold"
                            >
                              We are getting ready for your experience on IO
                            </Text>
                          </View>
                        </View>
                      </RNCSafeAreaView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
